### ISSUE TITLE
Allow skipping version metadata when specifying `--baseline-version`

### ISSUE THREAD

**Issue by maia-s** [obi1kenobi/cargo-semver-checks#1262]
### Steps to reproduce the bug with the above code

If a version of a crate has version metadata, that metadata has to be provided to `--baseline-version`, or it will error. For example, this fails:
```
> DOCS_RS=1 cargo semver-checks --baseline-version 0.4.7
    Building sdl3-sys v0.4.9+SDL3-3.2.10 (current)
       Built [   1.106s] (current)
     Parsing sdl3-sys v0.4.9+SDL3-3.2.10 (current)
      Parsed [   0.032s] (current)
error: Version sdl3-sys of crate 0.4.7 not found in registry
```
(also, the crate name and version are swapped in the error message)

But this succeeds:
```
> DOCS_RS=1 cargo semver-checks --baseline-version 0.4.7+SDL3-3.2.10
    Building sdl3-sys v0.4.9+SDL3-3.2.10 (current)
       Built [   1.431s] (current)
     Parsing sdl3-sys v0.4.9+SDL3-3.2.10 (current)
      Parsed [   0.033s] (current)
     Parsing sdl3-sys v0.4.7+SDL3-3.2.10 (baseline, cached)
      Parsed [   0.040s] (baseline)
    Checking sdl3-sys v0.4.7+SDL3-3.2.10 -> v0.4.9+SDL3-3.2.10 (minor change)
     Checked [   0.148s] 128 checks: 128 pass, 36 skip
     Summary no semver update required
    Finished [   2.160s] sdl3-sys
```

### Actual Behaviour

Version metadata is required

### Expected Behaviour

Version metadata is ignored by cargo, so `cargo-semver-checks` shouldn't require it.

### Generated System Information

System information:
-------------------
#### Software version

cargo-semver-checks 0.41.0 (793f25c)

#### Operating system

- OS: macOS 15.4.1 Sequoia
- Kernel: 24.4.0

#### Command-line

```bash
/Users/maia/.cargo/bin/cargo-semver-checks semver-checks --bugreport
```

#### cargo version

```
> cargo -V
cargo 1.88.0-nightly (d811228b1 2025-04-15)
```

#### Compile time information

- Profile: release
- Target triple: aarch64-apple-darwin
- Family: unix
- OS: macos
- Architecture: aarch64
- Pointer width: 64
- Endian: little
- CPU features: aes,crc,dit,dotprod,dpb,dpb2,fcma,fhm,flagm,fp16,frintts,jsconv,lor,lse,neon,paca,pacg,pan,pmuv3,ras,rcpc,rcpc2,rdm,sb,sha2,sha3,ssbs,vh
- Host: aarch64-apple-darwin

**Comment by Frank-III** [comment]
> Version metadata is ignored by cargo, so cargo-semver-checks shouldn't require it.

I think this makes sense, we could use `v.cmp_precedence` to ignore metadata in comparison: 
https://github.com/obi1kenobi/cargo-semver-checks/blob/ec3bf4ade25a32e10089ced87423bb6c058160ba/src/rustdoc_gen.rs#L721

willing to PR and fix it

**Comment by obi1kenobi** [comment]
Sorry for the delay, I'm still working through all the issues that were opened while I was travelling for RustWeek.

I'm fixing the error message issue in #1280. Thanks for pointing it out.

As for needing the full version including metadata, I think that works as intended. For example, the versions crates.io lists are these and v0.4.7 isn't one of them: https://crates.io/crates/sdl3-sys/versions

I think it's important to have parity between what crates.io considers a version and what we consider one. Otherwise we risk confusion and divergences in behavior that will be maintenance hell.

It's possible I might be missing some reason why requiring the metadata is a bad idea. Please do let me know if you feel that is the case!

**Comment by maia-s** [comment]
Yes, I do. crates.io is a database. It doesn't strip metadata because showing metadata there is the whole point, and it doesn't show duplicate versions without metadata because that would be both redundant, misleading, and require extra work for tools using it (suggesting there's two different versions when there's only one, and tools would have to filter that out)

`cargo-semver-checks` is a cargo tool, so it should aim for parity with *cargo*. cargo doesn't require providing version metadata. In fact, the docs explicitly says *not* to use version metadata when specifying dependencies:

> [Version metadata](https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field), such as 1.0.0+21AF26D3, is ignored and should not be used in version requirements.

(quote from <https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#version-metadata>)

**Comment by obi1kenobi** [comment]
Thanks for the reply. Here are my thoughts, lmk what you think.

> Yes, I do. crates.io is a database. It doesn't strip metadata because showing metadata there is the whole point, and it doesn't show duplicate versions without metadata

Actually at least for me it shows versions with metadata all the time, even when there are no duplicates: https://crates.io/crates/sdl3-sys/versions

> cargo doesn't require providing version metadata. In fact, the docs explicitly says *not* to use version metadata when specifying dependencies:
> 
>> [Version metadata](https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field), such as 1.0.0+21AF26D3, is ignored and should not be used in version requirements.

Note that you refer to version _requirements_, and I don't. `--baseline-version` does not accept _requirements_ — giving it `^1` would fail too. It accepts a single concrete version to look up. Unlike cargo, it isn't aiming to perform dependency resolution — it's looking for a specific entry in the crates.io version database.

Say that the user specifies `--baseline-version 1.2.3` and there are multiple `1.2.3` versions with different metadata on crates.io. What do you think `cargo-semver-checks` should do then? This seems like a confusing and unnecessary situation that also requires extra work.

In particular, we can't even reasonably say that "versions without metadata are fine if they are unambiguous" — that would mean that historical `cargo-semver-checks` workloads are non-reproducible. A new release at the same version but with different metadata may have been published since the `cargo-semver-checks` run, causing what was a previously unambiguous resolution to break on the ambiguity — or worse, pick the wrong release! Note that ordering by time as a tie-breaker doesn't work either because of yanking.

I know it's not what you wanted to hear, but I remain unconvinced. I also remain open to having my mind changed if you still feel I have this backwards.

**Comment by maia-s** [comment]
> Actually at least for me it shows versions with metadata all the time, even when there are no duplicates: https://crates.io/crates/sdl3-sys/versions

Yes, that's what I meant. The purpose of version metadata is to show it in places like that.

> Say that the user specifies --baseline-version 1.2.3 and there are multiple 1.2.3 versions with different metadata on crates.io. What do you think cargo-semver-checks should do then? This seems like a confusing and unnecessary situation that also requires extra work.

This is impossible, because version metadata is ignored. You can't publish version 1.2.3 of a crate twice with different metadata.

**Comment by maia-s** [comment]
```
❯ cargo publish --allow-dirty --dry-run
    Updating crates.io index
warning: crate sdl3-sys@0.5.2+different-metadata already exists on crates.io index
(...)
```
(the version that exists on crates.io is 0.5.2+SDL3-3.2.16)

**Comment by obi1kenobi** [comment]
> You can't publish version 1.2.3 of a crate twice with different metadata.

I just want to double-check a nuance here: if a version is yanked, can a version with the same number and different metadata be published?

**Comment by maia-s** [comment]
No.

Log from adding version metadata to sdl3-main 0.5.3, which has been yanked, and publishing:
```
❯ cargo publish --allow-dirty
    Updating crates.io index
   Packaging sdl3-main v0.5.3+yanked (/Users/maia/Dev/public/sdl3-sys-rs/sdl3-main)
    Updating crates.io index
    Packaged 12 files, 98.5KiB (16.6KiB compressed)
   Verifying sdl3-main v0.5.3+yanked (/Users/maia/Dev/public/sdl3-sys-rs/sdl3-main)
  Downloaded sdl3-sys v0.5.2+SDL3-3.2.16
  Downloaded 1 crate (404.5KiB) in 0.78s
   Compiling sdl3-sys v0.5.2+SDL3-3.2.16
   Compiling sdl3-main-macros v0.1.1
   Compiling sdl3-main v0.5.3+yanked (/Users/maia/Dev/public/sdl3-sys-rs/target/package/sdl3-main-0.5.3+yanked)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 4.25s
   Uploading sdl3-main v0.5.3+yanked (/Users/maia/Dev/public/sdl3-sys-rs/sdl3-main)
error: failed to publish to registry at https://crates.io

Caused by:
  the remote server responded with an error (status 400 Bad Request): crate version `0.5.3` is already uploaded
```

**Comment by obi1kenobi** [comment]
Neat, TIL. Thanks for testing all this out.

Given that we can't have ambiguity I think it's reasonable to look into supporting a bare version without metadata.

Would happily merge a PR here. cc @Frank-III in case you're still interested in looking into this — no pressure though.
